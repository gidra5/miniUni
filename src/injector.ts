import { FileMap } from 'codespan-napi';

enum Injectable {
  FileMap = 'FileMap',
  RootDir = 'RootDir',
}

type InjectableType = {
  [Injectable.FileMap]: FileMap;
  [Injectable.RootDir]: string;
};

const registry = new Map<string, any>();

const register = <const T extends Injectable>(
  name: T,
  value: InjectableType[T]
) => {
  registry.set(name, value);
};

const inject = <const T extends Injectable>(name: T): InjectableType[T] => {
  if (!registry.has(name)) {
    throw new Error(`Missing injection entry for: ${name}`);
  }

  return registry.get(name);
};

// Register default injectables
register(Injectable.RootDir, process.cwd());
register(Injectable.FileMap, new FileMap());

export { register, inject, Injectable };
