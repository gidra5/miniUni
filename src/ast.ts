import { SystemError } from './error.js';
import { inject, Injectable, register } from './injector.js';
import { isPosition, Position } from './position.js';
import { Token } from './tokens.js';

export type Tree = {
  type: string;
  id: string;
  data: any;
  children: Tree[];
};
export type Precedence = [prefix: number | null, postfix: number | null];

export const OperatorNodeType = {
  IMPLICIT_PLACEHOLDER: 'implicit_placeholder',
  PLACEHOLDER: 'placeholder',
  NAME: 'name',
  NUMBER: 'number',
  STRING: 'string',
  ATOM: 'atom',

  ADD: 'add',
  PLUS: 'plus',
  SUB: 'subtract',
  MINUS: 'minus',
  DIV: '/',
  MULT: '*',
  MOD: '%',
  POW: 'pow',
  PARALLEL: 'parallel',
  SEND: 'send',
  LABEL: ':',
  TUPLE: ',',
  NOT_EQUAL: '!=',
  EQUAL: '==',
  DEEP_EQUAL: '===',
  DEEP_NOT_EQUAL: '!==',
  AND: 'and',
  OR: 'or',
  LESS: '<',
  LESS_EQUAL: '<=',
  APPLICATION: 'application',
  SEQUENCE: 'sequence',
  SEND_STATUS: '?<-',
  GREATER: '>',
  GREATER_EQUAL: '>=',
  IS: 'is',
  IN: 'in',
  BIND: '@',
  PIN: 'pin',

  RECEIVE: 'receive',
  DECLARE: ':=',
  ASSIGN: '=',
  NOT: 'not',
  PARENS: 'parens',
  INDEX: 'index',
  SQUARE_BRACKETS: 'square_brackets',
  BLOCK: 'block',
  FUNCTION: 'func',
  IF: 'if',
  IF_ELSE: 'if_else',
  WHILE: 'while',
  IMPORT: 'import',
  POST_INCREMENT: 'post_increment',
  POST_DECREMENT: 'post_decrement',
  DECREMENT: '--',
  INCREMENT: '++',
  EXPORT: 'export',
  RECEIVE_STATUS: '<-?',
  INC_ASSIGN: '+=',
  LOOP: 'loop',
  FOR: 'for',
  FORK: 'fork',
  ASYNC: 'async',
  AWAIT: 'await',
  MATCH: 'match',
  MATCH_CASE: 'match_case',
  INJECT: 'inject',
  MASK: 'mask',
  WITHOUT: 'without',
  CODE_LABEL: '::',
  PIPE: '|>',

  SPREAD: '...',
  MUTABLE: 'mut',
} as const;
export type OperatorNodeType =
  (typeof OperatorNodeType)[keyof typeof OperatorNodeType];

export const NodeType = {
  ERROR: 'error',

  ...OperatorNodeType,

  SCRIPT: 'script',
  MODULE: 'module',

  RECORD: 'object',
  LIKE: 'like',
  STRICT: 'strict',
} as const;
export type NodeType = (typeof NodeType)[keyof typeof NodeType];

type ScriptNode = {
  id: string;
  data: {};
  type: typeof NodeType.SCRIPT;
  children: [ExpressionNode];
};
export type ModuleNode = {
  id: string;
  data: {};
  type: typeof NodeType.MODULE;
  children: [
    ...(ImportNode | ErrorNode | DeclarationPatternNode)[],
    ...([ExportNode] | [])
  ];
};
export type DeclarationPatternNode = {
  id: string;
  data: {};
  type: typeof NodeType.DECLARE;
  children: [ExportPatternNode, ExpressionNode];
};
export type ImportNode = {
  id: string;
  data: { name: string };
  type: typeof NodeType.IMPORT;
  children: [PatternNode];
};
export type ExportNode = {
  id: string;
  data: {};
  type: typeof NodeType.EXPORT;
  children: [ExpressionNode];
};
type ExportPatternNode = PatternNode;

export type ExpressionNode =
  | ErrorNode
  | ImplicitPlaceholderNode
  | PlaceholderNode
  | NameNode
  | NumberNode
  | StringNode
  | AtomNode
  | BlockNode
  | SequenceNode
  | VariableNode
  | MatchNode
  | FunctionNode
  | TupleNode;
export type ErrorNode = {
  id: string;
  data: { cause: SystemError };
  type: typeof NodeType.ERROR;
  children: [Tree] | [];
};
type ImplicitPlaceholderNode = {
  id: string;
  data: {};
  type: typeof NodeType.IMPLICIT_PLACEHOLDER;
  children: [];
};
type PlaceholderNode = {
  id: string;
  data: {};
  type: typeof NodeType.PLACEHOLDER;
  children: [];
};
type NameNode = {
  id: string;
  data: { value: string | symbol };
  type: typeof NodeType.NAME;
  children: [];
};
type NumberNode = {
  id: string;
  data: { value: number };
  type: typeof NodeType.NUMBER;
  children: [];
};
type StringNode = {
  id: string;
  data: { value: string };
  type: typeof NodeType.STRING;
  children: [];
};
type AtomNode = {
  id: string;
  data: { name: string };
  type: typeof NodeType.ATOM;
  children: [];
};

type BlockNode = {
  id: string;
  data: {};
  type: typeof NodeType.BLOCK;
  children: [ExpressionNode];
};
type SequenceNode = {
  id: string;
  data: {};
  type: typeof NodeType.SEQUENCE;
  children: ExpressionNode[];
};
type VariableNode = {
  id: string;
  data: { assert: boolean };
  type: typeof NodeType.DECLARE;
  children: [MatchPatternNode, ExpressionNode];
};
type TupleNode = {
  id: string;
  data: {};
  type: typeof NodeType.TUPLE;
  children: ExpressionNode[];
};
type MatchNode = {
  id: string;
  data: {};
  type: typeof NodeType.MATCH;
  children: [ExpressionNode, ...MatchCaseNode[]];
};
type MatchCaseNode = {
  id: string;
  data: {};
  type: typeof NodeType.MATCH_CASE;
  children: [MatchPatternNode, ExpressionNode];
};
type FunctionNode = {
  id: string;
  data: { isTopFunction?: false };
  type: typeof NodeType.FUNCTION;
  children: [MatchPatternNode, ExpressionNode];
};

type PatternNode =
  | ExpressionNode
  | RecordPatternNode
  | TuplePatternNode
  | PlaceholderNode
  | NameNode
  | StringNode
  | NumberNode
  | AtomNode;
type PatternWithDefaultNode = {
  id: string;
  data: {};
  type: typeof NodeType.EQUAL;
  children: [MatchPatternNode, ExpressionNode];
};
type MatchPatternNode = MutablePatternNode | PatternNode;

type MutablePatternNode = {
  id: string;
  data: {};
  type: typeof NodeType.MUTABLE;
  children: [PatternWithDefaultNode];
};

type TuplePatternNode = {
  id: string;
  data: {};
  type: typeof NodeType.TUPLE;
  children: [
    ...([SpreadPatternNode] | []),
    ...PatternWithDefaultNode[],
    ...([SpreadPatternNode] | [])
  ];
};
type RecordPatternNode = {
  id: string;
  data: {};
  type: typeof NodeType.RECORD;
  children: [
    ...(RecordNameWithDefaultNode | NameNode | RecordRenamePatternNode)[],
    ...([SpreadPatternNode] | [])
  ];
};
type RecordNameWithDefaultNode = {
  id: string;
  data: {};
  type: typeof NodeType.LABEL;
  children: [NameNode, ExpressionNode];
};
type SpreadPatternNode = {
  id: string;
  data: {};
  type: typeof NodeType.SPREAD;
  children: [MatchPatternNode];
};
type RecordRenamePatternNode = {
  id: string;
  data: {};
  type: typeof NodeType.LABEL;
  children: [NameNode | ExpressionNode, PatternWithDefaultNode];
};

const nextId = () => {
  const id = inject(Injectable.ASTNodeNextId);
  register(Injectable.ASTNodeNextId, id + 1);
  return String(id);
};

enum Associativity {
  LEFT = 'left',
  RIGHT = 'right',
  LEFT_AND_RIGHT = 'both',
}

enum Fixity {
  PREFIX = 'prefix',
  POSTFIX = 'postfix',
  INFIX = 'infix',
  NONE = 'none',
}

const generatePrecedences = <T extends string>(
  precedenceList: [T, Fixity, Associativity?][]
) => {
  const precedences = {} as Record<T, Precedence>;

  // if two same operators are next to each other, which one will take precedence
  // left associative - left one will take precedence
  // right associative - right one will take precedence
  // associative - does not matter, can be grouped in any order
  const leftAssociative = (p: number): Precedence => [p, p + 1];
  const rightAssociative = (p: number): Precedence => [p + 1, p];
  const associative = (p: number): Precedence => [p, p];
  let precedenceCounter = 0;

  for (const [operator, fixity, associativity] of precedenceList) {
    precedenceCounter++;

    if (fixity === Fixity.PREFIX) {
      precedences[operator] = [null, precedenceCounter];
    } else if (fixity === Fixity.POSTFIX) {
      precedences[operator] = [precedenceCounter, null];
    } else if (fixity === Fixity.NONE) {
      precedences[operator] = [null, null];
    } else if (associativity === Associativity.LEFT_AND_RIGHT) {
      precedences[operator] = associative(precedenceCounter);
    } else if (associativity === Associativity.LEFT) {
      precedences[operator] = leftAssociative(precedenceCounter++);
    } else precedences[operator] = rightAssociative(precedenceCounter++);
  }

  return precedences;
};

// if two same operators are next to each other, which one will take precedence
// first come lower precedence operators
const exprPrecedenceList: [NodeType, Fixity, Associativity?][] = [
  [NodeType.SEQUENCE, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.FUNCTION, Fixity.PREFIX],
  [NodeType.FORK, Fixity.PREFIX],
  [NodeType.IF, Fixity.PREFIX],
  [NodeType.IF_ELSE, Fixity.PREFIX],
  [NodeType.LOOP, Fixity.PREFIX],
  [NodeType.WHILE, Fixity.PREFIX],
  [NodeType.FOR, Fixity.PREFIX],
  [NodeType.INJECT, Fixity.PREFIX],
  [NodeType.MASK, Fixity.PREFIX],
  [NodeType.WITHOUT, Fixity.PREFIX],

  [NodeType.DECLARE, Fixity.PREFIX],
  [NodeType.ASSIGN, Fixity.PREFIX],
  [NodeType.INC_ASSIGN, Fixity.PREFIX],

  [NodeType.PARALLEL, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.PIPE, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.TUPLE, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.LABEL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.SPREAD, Fixity.PREFIX],

  [NodeType.SEND, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.RECEIVE, Fixity.PREFIX],
  [NodeType.SEND_STATUS, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.RECEIVE_STATUS, Fixity.PREFIX],

  [NodeType.OR, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.AND, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.EQUAL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.NOT_EQUAL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.DEEP_EQUAL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.DEEP_NOT_EQUAL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.LESS, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.LESS_EQUAL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.GREATER, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.GREATER_EQUAL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.IS, Fixity.POSTFIX],
  [NodeType.IN, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.NOT, Fixity.PREFIX],

  [NodeType.ADD, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.SUB, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.MULT, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.DIV, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.MOD, Fixity.INFIX, Associativity.LEFT],
  [NodeType.POW, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.MINUS, Fixity.PREFIX],
  [NodeType.PLUS, Fixity.PREFIX],
  [NodeType.INCREMENT, Fixity.PREFIX],
  [NodeType.DECREMENT, Fixity.PREFIX],
  [NodeType.CODE_LABEL, Fixity.PREFIX],
  [NodeType.POST_INCREMENT, Fixity.POSTFIX],
  [NodeType.POST_DECREMENT, Fixity.POSTFIX],

  [NodeType.IMPORT, Fixity.PREFIX],
  [NodeType.EXPORT, Fixity.PREFIX],
  [NodeType.MUTABLE, Fixity.PREFIX],
  [NodeType.AWAIT, Fixity.PREFIX],
  [NodeType.APPLICATION, Fixity.INFIX, Associativity.LEFT],
  [NodeType.ASYNC, Fixity.PREFIX],
  [NodeType.INDEX, Fixity.POSTFIX],
] as const;

const patternPrecedenceList: [NodeType, Fixity, Associativity?][] = [
  [NodeType.TUPLE, Fixity.INFIX, Associativity.LEFT_AND_RIGHT],
  [NodeType.LABEL, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.SPREAD, Fixity.PREFIX],
  [NodeType.ASSIGN, Fixity.POSTFIX],
  [NodeType.MUTABLE, Fixity.PREFIX],
  [NodeType.LIKE, Fixity.PREFIX],
  [NodeType.STRICT, Fixity.PREFIX],
  [NodeType.EXPORT, Fixity.PREFIX],
  [NodeType.NOT, Fixity.PREFIX],
  [NodeType.INDEX, Fixity.POSTFIX],
  [NodeType.BIND, Fixity.INFIX, Associativity.RIGHT],
  [NodeType.ATOM, Fixity.PREFIX],
] as const;

const exprPrecedences = generatePrecedences(exprPrecedenceList);

const patternPrecedences = generatePrecedences(patternPrecedenceList);

export const getExprPrecedence = (operator: string): Precedence => {
  return exprPrecedences[operator] ?? [null, null];
};

export const getPatternPrecedence = (operator: string): Precedence => {
  return patternPrecedences[operator] ?? [null, null];
};

type NodeOptions = {
  data?: any;
  position?: Position;
  children?: Tree[];
};

export const node = (
  type: string,
  { data = {}, position, children = [] }: NodeOptions = {}
): Tree => {
  const id = nextId();
  if (position) inject(Injectable.ASTNodePositionMap).set(id, position);
  return { type, id, data, children };
};

export const error = (cause: SystemError, _node: Tree | Position): ErrorNode =>
  node(NodeType.ERROR, {
    data: { cause },
    children: 'type' in _node ? [_node] : [],
    position: isPosition(_node) ? _node : undefined,
  }) as ErrorNode;

export const implicitPlaceholder = (
  position: Position
): ImplicitPlaceholderNode =>
  node(NodeType.IMPLICIT_PLACEHOLDER, { position }) as ImplicitPlaceholderNode;

export const placeholder = (position: Position): PlaceholderNode =>
  node(NodeType.PLACEHOLDER, { position }) as PlaceholderNode;

export const name = (value: string | symbol, position: Position): NameNode =>
  node(NodeType.NAME, { data: { value }, position }) as NameNode;

export const number = (value: number, position: Position): NumberNode =>
  node(NodeType.NUMBER, { data: { value }, position }) as NumberNode;

export const string = (value: string, position: Position): StringNode =>
  node(NodeType.STRING, { data: { value }, position }) as StringNode;

export const token = (token: Token, position: Position) =>
  token.type === 'number'
    ? number(token.value, position)
    : token.type === 'string'
    ? string(token.value, position)
    : token.type === 'placeholder'
    ? placeholder(position)
    : token.type === 'error'
    ? error(token.cause, position)
    : name(token.src, position);

export const atom = (name: string): AtomNode =>
  node(NodeType.ATOM, { data: { name } }) as AtomNode;

export const module = (children: ModuleNode['children']): ModuleNode =>
  node(NodeType.MODULE, { children }) as ModuleNode;

export const script = (expr: ExpressionNode): ScriptNode =>
  node(NodeType.SCRIPT, {
    children: expr.type === NodeType.SEQUENCE ? expr.children : [expr],
  }) as ScriptNode;

export const block = (expr: Tree, position?: Position): BlockNode =>
  node(NodeType.BLOCK, { position, children: [expr] }) as BlockNode;

export const sequence = (children: Tree[]): SequenceNode =>
  node(NodeType.SEQUENCE, { children }) as SequenceNode;

export const fn = (
  pattern: Tree,
  body: Tree,
  {
    position,
    isTopFunction = true,
  }: { position?: Position; isTopFunction?: boolean } = {}
): FunctionNode => {
  const children = [pattern, body];
  const _node = node(NodeType.FUNCTION, { position, children }) as FunctionNode;
  if (!isTopFunction) _node.data.isTopFunction = isTopFunction;
  return _node;
};

export const tuple = (children: Tree[]): TupleNode =>
  node(NodeType.TUPLE, { children }) as TupleNode;

export const loop = (body: Tree, position?: Position) =>
  node(NodeType.LOOP, { children: [body], position });
export const ifElse = (
  condition: Tree,
  ifTrue: Tree,
  ifFalse: Tree,
  position?: Position
) =>
  node(NodeType.IF_ELSE, { children: [condition, ifTrue, ifFalse], position });
export const application = (fn: Tree, arg: Tree) =>
  node(NodeType.APPLICATION, { children: [fn, arg] });
