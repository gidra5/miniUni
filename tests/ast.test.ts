import { it, beforeEach, describe, expect } from 'vitest';
import { parseTokens } from '../src/tokens.ts';
import { parseScript } from '../src/parser.ts';
import { Injectable, register } from '../src/injector.ts';
import { FileMap } from 'codespan-napi';
import { Tree } from '../src/ast.ts';

beforeEach(() => {
  register(Injectable.FileMap, new FileMap());
  register(Injectable.ASTNodeNextId, 0);
  register(Injectable.ASTNodePrecedenceMap, new Map());
  register(Injectable.ASTNodePositionMap, new Map());
});

const testCase = (input: string) => {
  const tokens = parseTokens(input);
  const ast = parseScript(tokens);

  expect(clearIds(ast)).toMatchSnapshot();

  function clearIds(ast: Tree) {
    if (ast.children.length > 0) {
      ast.children.forEach(clearIds);
    }
    delete (ast as any).id;
    return ast;
  }
};

describe('advent of code 1 single file', () => {
  it('variable', () =>
    testCase(`
      // https://adventofcode.com/2023/day/1

      /* take first and last digit on line, concat into two-digit number
        * and sum all numbers in document
        */
      document := "
        1abc2
        pqr3stu8vwx
        a1b2c3d4e5f
        treb7uchet
      "
    `));

  it('split lines', () =>
    testCase(`
        lines := {
          lines := split document "\\n"
          lines = map lines (replace "\\w+" "")
          lines = filter lines fn line do line != ""
        }
      `));

  it('parse numbers', () =>
    testCase(`
        numbers := flat_map lines fn line {
          digits := ()
  
          while line != "" {
            if match "\d" (char_at line 0) {
              digit := number(char_at line 0)
              if !digits[0] do digits[0] = digit
              digits[1] = digit
            }
            (_, ...line) = line
          }
  
          digits[0], digits[1] * 10
        }
      `));

  it('fn multiple args', () => testCase(`fn acc, item -> ()`));

  it('flat map list reducer', () =>
    testCase(`fn acc, item -> (...acc, ...mapper item)`));

  it('flat map list impl', () =>
    testCase(`
        flat_map := fn list, mapper {
          reduce list (fn acc, item -> (...acc, ...mapper item)) (fn first, second -> (...first, ...second)) ()
        }
      `));

  it('reduce list', () =>
    testCase(`
        reduce := fn list, reducer, merge, initial {
          len := length list
          if len == 0 do return initial
        
          midpoint := floor(len / 2)
          item := list[midpoint]
          first, second := all(
            | (reduce slice(list, 0, midpoint) reducer merge initial)
            | (reduce slice(list, midpoint + 1) reducer merge initial)
          )
        
          merge (reducer first item) second
        }
      `));
});

/* one test per example of a language construct  */

describe('comments', () => {
  it('comment', () => testCase(`// comment\n123`));

  it('comment block', () => testCase(`/* comment block */123`));
});

describe('expressions', () => {
  describe('values', () => {
    it('integer', () => testCase(`123`));
    it('float', () => testCase(`123.456`));
    it('string', () => testCase(`"string"`));
  });

  describe('arithmetics', () => {
    it('order of application', () => testCase(`1 + 2^-3 * 4 - 5 / 6 % 7`));
    it('-(a+b)', () => testCase(`-(a+b)`));

    it('complex', () =>
      testCase(
        `(2^2-5+7)-(-i)+ (j)/0 - 1*(1*f)+(27-x )/q + send(-(2+7)/A,j, i, 127.0 ) + 1/1`
      ));
  });

  describe('boolean expressions', () => {
    it('not', () => testCase(`!123`));
    it('"and", "or" and "not" associativity', () =>
      testCase(`a and b and c or d or !e and f and g or not h or i`));
    it('in operator', () => testCase(`:key in x and y`));
  });

  describe('function expressions', () => {
    it('function block body', () => testCase(`fn x, y { x + y }`));
    it('function multiple params', () => testCase(`fn x, y -> x + y`));
    it('fn no parameters', () => testCase(`fn -> 123`));
    it('fn no parameters block', () => testCase(`fn { 123 }`));
    it('arrow function', () => testCase(`x -> x`));
    it('fn increment', () => testCase(`fn -> line_handled_count++`));

    describe('application', () => {
      it('function call', () => testCase(`f x`));
      it('function call multiple args', () => testCase(`f x y`));
      it('function call param with field', () => testCase(`f x.y`));
      it('send((1+2), 3)', () => testCase(`send((1+2), 3)`));
      it('send(2, 3)', () => testCase(`send(2, 3)`));
      it('(send)(2, 3)', () => testCase(`(send)(2, 3)`));
      it('(send 1)(2, 3)', () => testCase(`(send 1)(2, 3)`));
      it('(send 1 2)(2, 3)', () => testCase(`(send 1 2)(2, 3)`));
      it('send 1 + 2', () => testCase(`send 1 + 2`));
      it('a + send (2, 3)', () => testCase(`a + send (2, 3)`));
      it('send a (2, 3)', () => testCase(`send a (2, 3)`));
      it('send 1 (2, 3)', () => testCase(`send 1 (2, 3)`));
      it('a + send 1 + 2', () => testCase(`a + send 1 + 2`));
      it('methods chaining', () => testCase(`math.floor(1).multiply(2)`));
      it('function as last arg', () =>
        testCase(`open "file" file -> write file "yolo"`));
    });

    describe('function forms', () => {
      it('immediate form', () => testCase(`fn do x; y`));
      it('block form', () => testCase(`fn { x }`));
      it('rest form', () => testCase(`fn -> x; y`));
    });
  });

  describe('pattern matching', () => {
    it('switch', () => testCase(`switch a { 1 -> 2, 2 -> 3, _ -> 4 }`));

    it('in function parameters', () => testCase(`(x, y) -> x + y`));
    it('declare record pattern', () => testCase(`{ a, b } := handlers`));
    it("with 'is' operator", () => testCase(`x is (a, b)`));
    it('with placeholder', () => testCase(`x is (_, b)`));
    it('with pin', () => testCase(`x is (^a, b)`));
    it('with pin expression', () => testCase(`x is (^(a + b), b)`));
    it('with constant value', () => testCase(`x is (1, b)`));
    it('with rest value', () => testCase(`x is (a, ...b)`));
    it('with rest value first', () => testCase(`x is (...b, a)`));
    it('with record pattern', () => testCase(`x is { a, b }`));
    it('with record pattern rename', () => testCase(`x is { a: c, b }`));
    it('with record pattern key', () => testCase(`x is { [a + b]: c, b }`));
    it('with record pattern nested', () => testCase(`x is { a: (c, d), b }`));
    it('with default value', () => testCase(`x is ((b = 4), a)`));
    it.todo('with record default value', () => testCase(`x is { b = 4, a }`));
    it('with rename', () => testCase(`x is (a @ b, c)`));
    it('with name for match', () => testCase(`x is ((a, b) @ c)`));

    it('binding visible in scope where it is true', () =>
      testCase(`x is (a, b) and a == b + 1`));
  });

  describe('structured programming', () => {
    it('complex 1', () =>
      testCase(`
        y := {
          x := 25
          loop if x < 0 do break x else {
            y := x
            x = x - 1
            if y == 19 do continue 69
            y
          }
        }
      `));

    it('if-then', () => testCase(`if true do 123`));
    it('if-then-else', () => testCase(`if true do 123 else 456`));
    it('if-then-elseif-then-else', () =>
      testCase(`if true do 123 else if false do 789 else 456`));
    it('sequence', () => testCase(`123; 234; 345; 456`));
    it('block sequence', () => testCase(`{ 123; 234; 345; 456 }`));
    it('parens sequence', () => testCase(`(123; 234; 345; 456)`));
    it('block', () => testCase(`{ 123 }`));
    it('for loop', () => testCase(`for x in (1, 2, 3) do x`));
    it('while loop', () => testCase(`while true do 123`));
    it('loop', () => testCase(`loop 123`));
    it('loop scope', () => testCase(`loop { x }`));
    it('labeled expression', () => testCase(`label::123`));
    it('semicolon at the end', () => testCase(`1;`));

    describe('statement forms', () => {
      it('immediate form', () => testCase(`if true do 123; 456`));
      it('block form', () => testCase(`if true { 123 }`));
      it('rest form', () => testCase(`if true -> 123; 456`));
    });
  });

  describe('concurrent programming', () => {
    it('channel send', () => testCase(`c <- 123`));
    it('channel receive', () => testCase(`<- c`));
    it('channel try send', () => testCase(`c ?<- 123`));
    it('channel try receive', () => testCase(`<-? c`));
    it('try receive with assignment', () => testCase(`status := <-?numbers`));
    it('parallel value', () => testCase(`123 | 456`));
    it('prefix parallel with code after', () =>
      testCase(`| { };numbers := channel()`));
    it('parallel with channels', () => testCase(`c <- 123 | <- c`));
    it('fork', () => testCase(`fork f x; y`));
    it('async', () => testCase(`async f x`));
    it('async index', () => testCase(`async f.a`));
    it('await async', () => testCase(`await async f x`));
    it('await', () => testCase(`await x + 1`));
  });

  describe('data structures', () => {
    it('unit', () => testCase(`()`));
    it('tuple', () => testCase(`list, reducer, merge, initial`));
    it('record single', () => testCase(`a: 1`));
    it('record', () => testCase(`a: 1, b: 2`));
    it('tuple with single item (atom)', () => testCase(`(:a,)`));
    it('(-(2+7)/A,j, i, 127.0 )', () => testCase(`(-(2+7)/A,j, i, 127.0 )`));
    it('atom (global symbol)', () => testCase(`:atom`));
    it('dictionary', () => testCase(`[1]: 2, [3]: 4`));
    it('map without braces', () => testCase(`1+2: 3, 4+5: 6`));
    it('period operator', () => testCase(`math.floor`));
    it('index', () => testCase(`x[0]`));
    it('field assignment', () => testCase(`x.y = 123`));
    it('field assignment dynamic', () => testCase(`x[y] = 123`));
  });

  describe('effect handlers', () => {
    it('inject', () => testCase(`inject a: 1, b: 2 { 1 }`));
    it('mask', () => testCase(`mask "a", "b" { 1 }`));
    it('without', () => testCase(`without "a", "b" { 1 }`));
    it('complex', () =>
      testCase(`
        inject a: 1, b: 2 {
          { a, b } := handlers
          inject a: a+1, b: b+2 {
            mask "a" {
              without "b" {
                { a } := handlers
                a + 1
              }
            }
          }  
        }
      `));
  });
});

describe('programs', () => {
  it('import', () => testCase(`import "a" as b`));

  describe('script', () => {
    it('dynamic import', () => testCase(`b := import "a"`));
    it('dynamic async import', () => testCase(`b := async import "a"`));
  });

  describe('module', () => {
    it('export declaration', () => testCase(`export x := 123`));
    it('export default', () => testCase(`export fn args -> 1`));
  });
});

describe('newline handling', () => {
  it('for loop newline', () => testCase(`for x in 1, 2, 3 do\n x`));
  it('parallel parens', () => testCase(`(\n| 1\n| 2\n)`));
  it('parallel', () => testCase(`| 1\n| 2`));
  it('chaining', () => testCase(`a\n.b`));
  it('parens', () => testCase(`(\n1 +\n2\n+ 3\n)`));
  it('no parens', () => testCase(`1 +\n2\n+ 3`));
  it('prefix', () => testCase(`!\na`));
  it('infix-prefix', () => testCase(`b :=\n !\na`));
  it('infix-infix', () => testCase(`b +\nc +\nd`));
  it('if else separate lines', () => testCase(`if a do\n 1\n else\n 2`));
  it('if-then newline', () => testCase(`if true do\n 123`));
  it('if-then newline-else', () => testCase(`if true do\n 123 else 456`));
  it('if-then newline-else newline', () =>
    testCase(`if true do\n 123 else\n 456`));
  it('block newline in the middle', () => testCase(`{ a := 1\n b := 2 }`));
  it('block newline at the end', () => testCase(`{ a := 1\n b := 2\n }`));
  it('block newline at the beginning', () => testCase(`{\n a := 1\n b := 2 }`));
  it('block semicolon newline', () => testCase(`{ a := 1;\n b := 2 }`));
  it('block semicolon newline at the end', () =>
    testCase(`{ a := 1;\n b := 2;\n }`));
  it('variable', () => testCase(`1\n`));
  it('empty switch with newline', () => testCase(`switch a { \n }`));
});
