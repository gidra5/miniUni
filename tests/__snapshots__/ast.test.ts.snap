// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ast 1 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "data": {
                "value": "document",
              },
              "name": "name",
            },
          ],
          "data": {
            "operator": "pattern",
          },
          "name": "operator",
        },
        {
          "children": [],
          "data": {
            "value": "
        1abc2
        pqr3stu8vwx
        a1b2c3d4e5f
        treb7uchet
      ",
          },
          "name": "string",
        },
      ],
      "data": {
        "operator": ":=",
      },
      "name": "operator",
    },
  ],
  "data": {},
  "name": "script",
}
`;
